rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,219 @@
+    /// returns SlateDbError::InvalidArgument if `lower_bound` is `Unbounded`
+    /// returns SlateDbError::InvalidArgument if the key is comes before the
+    ///  current iterator position.
+    /// returns Err(InvalidatedIterator) if the 
+    pub async fn seek(
+        &mut self,
+        lower_bound: Bound<&[u8]>
+    ) -> Result<(), SlateDbError>;
+}
+
+struct DbRecord {
+    key: &[u8],
+    value: &[u8],
+}
+```
+
+Note that the `DbIterator` may be invalidated internally if any of its referenced
+resources must be reclaimed. In this case, a call to `DbIterator::next` or 
+`DbIterator::scan` will return `SlateDb::InvalidatedError`. This is discussed in 
+more detail below.
+
+We will also provide `Db::scan_with_options`, which allows the user to specify the
+query isolation through a `ScanOptions` struct. Similar to `get_with_options`, the
+two isolation levels that will be supported are `Committed` and `Uncommitted`.
+The default when `scan` is called without explicit options will be `Committed`.
+
+```rust
+impl Db {
+    /// Scan a range of keys with the provided options.
+    ///
+    /// returns a `DbIterator`  
+    pub async fn scan_with_options(
+        &self,
+        range: &DbRange,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
+}
+```
+
+## Implementation
+
+A scan must iterate every table which may have keys within the query range.
+This includes the active WAL, memtables waiting to be flushed, and L0/SR tables 
+which may need to be loaded from object storage. Internally, we will rely on 
+iterators which sort and merge these tables.  The latest records will
+have precedence following the same rules that a normal `get` query follows.
+
+The existing `MergeIterator` used for compaction provides much of the necessary
+foundation to unify multiple iterators. Additionally, `MemTableIterator` and 
+`SstIterator` provide ready-to-use capabilities to scan tables which may keys
+within the range. 
+
+**Query Isolation**: When using the `Uncommitted` isolation, unflushed
+records recorded in WALs will be used as a source for the key range. Otherwise,
+only committed records in memtables, as well as L0 and SR tables will be used.
+
+Range queries must access mutable table state, such as the current WAL. It is 
+therefore possible for the query results to include some records which 
+were written after the  query began scanning its range of records. For example,
+a query may include  only a subset of records that were inserted in the same 
+`WriteBatch`. In  other words, range queries will not have snapshot isolation 
+initially. 
+
+[RFC-????](https://github.com/slatedb/slatedb/pull/260) proposes transactional
```

Comment:
👍 Was literally just writing this in the review comment.

@flaneur2020 I don't _think_ this should have too much conflict with the transaction/SI/SSI design, right? Seems to me the iterator will just need to grab a seqnum when it starts and filter out all kv pairs > the seqnum. Shouldn't require any interaction with the oracle.

Reply:
sorry for late response

i think so, all it need is letting the iterator which filters out the rows with higher seqnum, the logic would be pretty trivial 🤔